src/libbson/src/bson/bson-dsl.md

```diff
@@ -0,0 +1,974 @@
+(DSL) built upon the C99 preprocessor. The DSL emits valid C99 and C++03 code
+that can be used to inspect and construct `bson_t` objects.
+
+The intent of the DSL is to simplify and streamline the process of working with
+BSON data within C, which is otherwise slow, error-prone, and requires a lot of
+boilerplate. The DSL presents a declarative style of working with BSON data.
+
+The DSL can be broken down into two classes:
+
+- [BSON writing](#generating-bson-data):
+  - [`bsonBuild`](#bsonbuild)
+  - [`bsonBuildDecl`](#bsonbuilddecl)
+  - [`bsonBuildArray`](#bsonbuildarray)
+  - [`bsonBuildAppend`](#bsonbuildappend)
+- [BSON reading](#reading-bson-data)
+  - [`bsonVisitEach`](#bsonvisiteach)
+  - [`bsonParse`](#bsonparse)
+
+The following DSL "grammatical elements" are defined:
+
+- [*DocOperation*](#docoperation)
+- [*ValueOperation*](#valueoperation)
+- [*ArrayOperation*](#arrayoperation)
+- [*ParseOperation*](#parseoperation)
+- [*VisitOperation*](#visitoperation)
+- [*Predicate*](#predicate)
+
+**NOTE:** The `bson-dsl.h` header is *not* a public API. It is intended for
+internal use and may be changed, broken, updated, or removed.
+
+# Generating BSON Data
+
+## Examples
+
+### Empty
+
+```c
+bsonBuildDecl(empty);
+```
+
+Result:
+
+```js
+{}
+```
+
+### A Simple "okay"
+
+```c
+bsonBuildDecl(e, kv("okay", bool(true)));
+```
+
+Result:
+
+```js
+{ okay: true }
+```
+
+### Nested Documents and Arrays
+
+```c
+bsonBuildDecl(e,
+    kv("integer", int32(8)),
+    kv("subdoc", doc(kv("foo", cstr("bar")), kv("baz", null))),
+    kv("anArray", array(int32(1), int32(2), null, cstr("I am a string"))));
+```
+
+Result:
+
+```js
+{
+  integer: 8,
+  subdoc: { foo: "bar", "baz": null },
+  anArray: [1, 2, null, "I am a string"]
+}
+```
+
+### Create an Aggregate Command
+
+```c
+bson_t* make_aggregate(const char* coll,
+                       const bson_t* pipeline_array,
+                       int timeout_ms,
+                       int batch_size,
+                       const bson_t* comment,
+                       bson_error_t* error) {
+  bson_t* r = bson_new();
+  bsonBuildAppend(r,
+    kv("aggregate", cstr(coll)),
+    kv("pipeline", bsonArray(*pipeline_array)),
+    kv("cursor",
+       doc(if(batch_size != -1,
+              then(kv("batchSize", int32(batch_size)))))),
+    if(timeout_ms > 0, then(kv("maxTimeMS", int32(timeout_ms))),
+    if(comment, then(kv("comment", bson(*comment)))));
+  // Check if any of the above failed:
+  if (bsonBuildError) {
+    bson_set_error(error,
+                   ERROR_COND,
+                   ERROR_CODE,
+                   "Failed to build aggregate: %s",
+                   bsonBuildError);
+    bson_destroy(r);
+    r = NULL;
+  }
+  return r;
+}
+```
+
+
+## `bsonBuild`
+
+> `bsonBuild(BSON, DocOperation... ops)`
+
+Calls `bson_init(&BSON)`, followed by `bsonBuildAppend(BSON, ops...)`. If the
+build operation results in an error, calls `bson_destroy(&BSON)`.
```

Comment:
Aside: In an earlier revision, I attempted to build the document into a temporary, then move it into the destination after it succeeded, otherwise it would just destroy the temporary and leave the destination unchanged. This had the added benefit that the destination document could still be used during construction. This proved infeasible as there was no API to "move" the ownership of the document into another `bson_t` (The intuitive API `bson_steal` marks the destination as read-only for some reason?)

Reply:
> The intuitive API bson_steal marks the destination as read-only for some reason?

That is a rough edge. `bson_steal` sets the destination `bson_t` flag `BSON_FLAG_STATIC`. `BSON_FLAG_STATIC` means the `bson_t` struct is not freed on `bson_destroy`. It is independent of the underlying data.

The caveat is `bson_steal` cannot be used on a `bson_t` that has been heap allocated. E.g.

```
bson_t *b = bson_new ();
bson_steal (b, src); // leaks the bson_t struct
```